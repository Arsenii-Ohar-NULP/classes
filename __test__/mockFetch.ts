export const mockFetch = () => {
    global.fetch = jest.fn();
}