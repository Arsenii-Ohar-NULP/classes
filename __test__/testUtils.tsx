import React, { PropsWithChildren } from 'react';
import { render, renderHook } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import type { RootState, AppStore } from 'pages/redux/store';
// As a basic setup, import your same slice reducers
import { authInitialState, authSlice } from 'pages/redux/auth';
import { classSlice, classesInitialState } from 'pages/redux/classes';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderHookWithProviders(
  ui,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...renderHook(() => ui(), { wrapper: Wrapper, ...renderOptions }) };
}
