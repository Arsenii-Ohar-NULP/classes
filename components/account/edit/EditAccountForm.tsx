import React, {useState} from 'react';
import EditAccountInput from 'components/account/edit/EditAccountInput';
import {useAppDispatch, useAppSelector} from 'components/redux/store';
import EditButton from 'components/account/edit/EditButton';
import {logout} from 'components/login/AuthService';
import {useRouter} from 'next/navigation';
import {authActions} from 'components/redux/auth';
import {useEditAccountForm} from "components/account/edit/useEditAccountForm";
import {EditUserData, useEditUserMutation} from "components/redux/userApi";
import User from "components/account/User";

interface ChangedData {
    Password: string;
    Email: string;
    Phone: string;
}

export default function EditAccountForm() {
    const dispatch = useAppDispatch();
    const router = useRouter();
    const user = useAppSelector((state) => state.auth.user);
    const [error, setError] = useState('');
    const [editUser, response] = useEditUserMutation();

    const {
        register,
        handleSubmit,
        formState: {errors},
        watch
    } = useEditAccountForm();

    const formData = {
        Password: watch('Password', '') as string,
        Email: watch('Email', user?.email ? user.email : '') as string,
        Phone: watch('Phone', user?.phone ? user.phone : '') as string
    };

    const findDifference = (obj1: ChangedData, obj2: User) => {
        if (!obj1 || !obj2) {
            return null;
        }
        const changedData: EditUserData = {id: obj2?.id};
        const keys: Array<keyof ChangedData> = Object.keys(obj1) as Array<keyof ChangedData>;
        for (const key of keys) {
            if (obj2[key.toLowerCase()] !== obj1[key] && obj1[key] !== '') {
                changedData[key.toLowerCase()] = obj1[key];
            }
        }

        return changedData;
    };

    const getUserChanges = () => {
        const userDifference = findDifference(formData, user);
        return userDifference && hasKeysNumber(userDifference, 1) ? userDifference : null;
    };

    const hasKeysNumber = (data: object, target: number) => {
        return Object.keys(data).length !== target;
    };

    const onSubmit = () => {
        const userChanges = getUserChanges();
        if (!userChanges) {
            alert('Data has not been changed');
            return;
        }

        editUser(userChanges).unwrap()
            .then(() => {
                dispatch(authActions.updateUser(userChanges));
                router.push('/main/account');
            })
            .catch((error) => {
                if ("status" in error) {
                    if (error['status'] === 401) {
                        logout(dispatch, router);
                    }
                }
                setError(error.message);
            })
    };

    return (
        <div className="d-flex align-items-center flex-column container">
            <h3 className="p-3">Edit account information</h3>
            <form>
                <EditAccountInput
                    type={'password'}
                    id={'Password'}
                    placeholder="Enter a new password"
                    errorMessage={errors?.Password?.message.toString()}
                    registration={register('Password')}
                />
                <EditAccountInput
                    id={'Phone'}
                    placeholder="Enter a new phone number"
                    errorMessage={errors?.Phone?.message.toString()}
                    value={user?.phone}
                    registration={register('Phone')}
                />
                <EditAccountInput
                    id={'Email'}
                    placeholder="Enter an email"
                    errorMessage={errors?.Email?.message.toString()}
                    type={'email'}
                    value={user?.email}
                    registration={register('Email')}
                />
                <hr/>
                <h6>{error}</h6>
                <EditButton
                    onClick={handleSubmit(onSubmit)}
                    disabled={!getUserChanges() || response.isLoading}
                />
            </form>
        </div>
    );
}
