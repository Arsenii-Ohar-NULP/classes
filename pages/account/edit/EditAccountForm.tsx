import React, { useState } from 'react';
import * as yup from 'yup';
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import { EditAccountInput } from './EditAccountInput';
import { useAppDispatch, useAppSelector } from 'pages/redux/store';
import EditButton from './EditButton';
import { editUser } from 'pages/account/UserService';
import InvalidCredentials from 'pages/errors/InvalidCredentials';
import { logout } from 'pages/login/AuthService';
import { useRouter } from 'next/router';
import { authActions } from 'pages/redux/auth';

interface IChangedData {
  Password: string;
  Email: string;
  Phone: string;
}

export default function EditAccountForm() {
  const dispatch = useAppDispatch();
  const router = useRouter();
  const user = useAppSelector((state) => state.auth.user);
  const [error, setError] = useState('');

  const schema = yup
    .object({
      Password: yup.string().notRequired(),
      Phone: yup.string().notRequired().min(8),
      Email: yup.string().email().notRequired(),
    })
    .required();

  const {
    register,
    handleSubmit,
    formState: { errors },
    watch
  } = useForm({
    resolver: yupResolver(schema),
  });
  const password = watch('Password', '');
  const email = watch('Email', user?.email ? user.email : '');
  const phone = watch('Phone', user?.phone ? user.phone : '');

  const cleanChangedData = (data: IChangedData) => {
    const changedData = {};
    if (!user){
      return changedData;
    }
    for (const key of Object.keys(data)) {
      if (user[key.toLowerCase()] !== data[key] && data[key] !== '') {
        changedData[key.toLowerCase()] = data[key];
      }
    }

    return changedData;
  };

  const hasDataChanged = (data) => {
    console.log(Object.keys(data).length !== 0);
    return Object.keys(data).length !== 0;
  };

  const onSubmit = (data: {
    Password: string;
    Phone: string;
    Email: string;
  }) => {
    const changedData = cleanChangedData(data);
    changedData['id'] = user.id;

    if (!hasDataChanged(changedData)) {
      alert('Data has not been changed');
      return;
    }

    editUser(changedData)
      .then(() => {
        alert('Successfully edited a user');
        dispatch(authActions.updateUser(changedData));
      })
      .catch((error) => {
        if (error instanceof InvalidCredentials) {
          logout(dispatch, router);
        }
        setError(error.message);
      });
  };

  return (
    <div className="d-flex align-items-center flex-column container">
      <h3 className="p-3">Edit account information</h3>
      <form>
        <EditAccountInput
          type={'password'}
          id={'Password'}
          placeholder="Enter a new password"
          errorMessage={errors?.Password?.message.toString()}
          registration={register('Password')}
        />
        <EditAccountInput
          id={'Phone'}
          placeholder="Enter a new phone number"
          errorMessage={errors?.Phone?.message.toString()}
          value={user?.phone}
          registration={register('Phone')}
        />
        <EditAccountInput
          id={'Email'}
          placeholder="Enter an email"
          errorMessage={errors?.Email?.message.toString()}
          type={'email'}
          value={user?.email}
          registration={register('Email')}
        />
        <hr />
        <h6>{error}</h6>
        <EditButton
          onClick={handleSubmit(onSubmit)}
          disabled={!hasDataChanged(cleanChangedData({
            Email: email,
            Phone: phone,
            Password: password
          }))}
        />
      </form>
    </div>
  );
}
