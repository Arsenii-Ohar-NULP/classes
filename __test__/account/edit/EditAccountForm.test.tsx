import React from 'react'
import { renderWithProviders } from '__test__/testUtils'
import EditAccountForm from 'pages/account/edit/EditAccountForm';
jest.mock('pages/login/AuthService');
jest.mock('pages/account/UserService');
jest.mock('next/router', () => ({ useRouter: jest.fn() }));

describe('edit account form', () => {
    it('should match a snapshot', () => {
        const page = renderWithProviders(<EditAccountForm/>);
        expect(page).toMatchSnapshot();
    })
})