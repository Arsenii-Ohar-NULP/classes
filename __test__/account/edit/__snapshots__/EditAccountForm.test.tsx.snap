// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit account form should match a snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="d-flex align-items-center flex-column container"
      >
        <h3
          class="p-3"
        >
          Edit account information
        </h3>
        <form>
          <div
            class="my-2 has-validation"
          >
            <label
              class="form-label"
              for="Password"
            >
              Password
            </label>
            <input
              class="form-control rounded p-1 px-2"
              id="Password"
              name="Password"
              placeholder="Enter a new password"
              type="password"
              value=""
            />
            <div
              class="invalid"
            />
          </div>
          <div
            class="my-2 has-validation"
          >
            <label
              class="form-label"
              for="Phone"
            >
              Phone
            </label>
            <input
              class="form-control rounded p-1 px-2"
              id="Phone"
              name="Phone"
              placeholder="Enter a new phone number"
              value=""
            />
            <div
              class="invalid"
            />
          </div>
          <div
            class="my-2 has-validation"
          >
            <label
              class="form-label"
              for="Email"
            >
              Email
            </label>
            <input
              class="form-control rounded p-1 px-2"
              id="Email"
              name="Email"
              placeholder="Enter an email"
              type="email"
              value=""
            />
            <div
              class="invalid"
            />
          </div>
          <hr />
          <h6 />
          <div
            class="text-center"
          >
            <button
              class="btn btn-primary"
              disabled=""
              type="submit"
            >
              Save
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="d-flex align-items-center flex-column container"
    >
      <h3
        class="p-3"
      >
        Edit account information
      </h3>
      <form>
        <div
          class="my-2 has-validation"
        >
          <label
            class="form-label"
            for="Password"
          >
            Password
          </label>
          <input
            class="form-control rounded p-1 px-2"
            id="Password"
            name="Password"
            placeholder="Enter a new password"
            type="password"
            value=""
          />
          <div
            class="invalid"
          />
        </div>
        <div
          class="my-2 has-validation"
        >
          <label
            class="form-label"
            for="Phone"
          >
            Phone
          </label>
          <input
            class="form-control rounded p-1 px-2"
            id="Phone"
            name="Phone"
            placeholder="Enter a new phone number"
            value=""
          />
          <div
            class="invalid"
          />
        </div>
        <div
          class="my-2 has-validation"
        >
          <label
            class="form-label"
            for="Email"
          >
            Email
          </label>
          <input
            class="form-control rounded p-1 px-2"
            id="Email"
            name="Email"
            placeholder="Enter an email"
            type="email"
            value=""
          />
          <div
            class="invalid"
          />
        </div>
        <hr />
        <h6 />
        <div
          class="text-center"
        >
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Save
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
